README.md
<!-- 
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages). 

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages). 
-->

# json_data
Helper package for converting nested json

## Features
* string -> json array (List)
* string -> json object (Map)
* List -> stringified json
* Map -> stringified json

## Getting started

* No prerequisites required

## Usage

```dart
// return : stringified json
JsonData.toJsonString({
      'hello': 'world',
      'list': [
        1,
        null,
        true,
        'flutter',
        false,
        [456, 654],
        {
          'deep-level': null,
          'deeper-level': {'hello': 'world'}
        }
      ]
    });
```
```dart
// return : Map<String, dynamic>
JsonData.fromJsonString(
        '{"hello": "world", "list": [1, null, true, "flutter", false, [456, 654], {"deep-level": null, "deeper-level": {"hello": "world"}}]}')
```

## Additional information
* If you have any issue or suggestion contact me via github.
[github repository](https://github.com/swimmingkiim/json_data)


CHANGELOG.md
## 0.0.1
initialize json_data package
* toJsonString
* toJsonArrayString
* fromJsonString
* fromJsonArrayString


test/json_data_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:json_data/json_data.dart';

void main() {
  group('toJsonString', () {
    test('simple map object', () {
      expect(JsonData.toJsonString({'hello': 'world'}), '{"hello": "world"}');
    });
    test('complex map object', () {
      expect(
        JsonData.toJsonString({
          'hello': 'world',
          'list': [
            1,
            null,
            true,
            'flutter',
            false,
            [456, 654],
            {
              'deep-level': null,
              'deeper-level': {'hello': 'world'}
            }
          ]
        }),
        '{"hello": "world", "list": [1, null, true, "flutter", false, [456, 654], {"deep-level": null, "deeper-level": {"hello": "world"}}]}',
      );
    });
  });

  group('toJsonArrayString', () {
    test('simple list object', () {
      expect(
          JsonData.toJsonArrayString([
            {'hello': 'world'}
          ]),
          '[{"hello": "world"}]');
    });
    test('complex list object', () {
      expect(
        JsonData.toJsonArrayString([
          {
            'hello': 'world',
            'list': [
              1,
              null,
              true,
              'flutter',
              false,
              [456, 654],
              {
                'deep-level': null,
                'deeper-level': {'hello': 'world'}
              }
            ]
          }
        ]),
        '[{"hello": "world", "list": [1, null, true, "flutter", false, [456, 654], {"deep-level": null, "deeper-level": {"hello": "world"}}]}]',
      );
    });
  });

  group('fromJsonString', () {
    test('simple json', () {
      expect(JsonData.fromJsonString('{"hello": "world"}'), {'hello': 'world'});
    });
    test('complex json', () {
      expect(
          JsonData.fromJsonString(
              '{"hello": "world", "list": [1, null, true, "flutter", false, [456, 654], {"deep-level": null, "deeper-level": {"hello": "world"}}]}'),
          {
            'hello': 'world',
            'list': [
              1,
              null,
              true,
              'flutter',
              false,
              [456, 654],
              {
                'deep-level': null,
                'deeper-level': {'hello': 'world'}
              }
            ]
          });
    });
  });
  group('fromJsonArrayString', () {
    test('simple json array', () {
      expect(JsonData.fromJsonArrayString('[{"hello": "world"}]'), [
        {'hello': 'world'}
      ]);
    });
    test('complex json array', () {
      expect(
          JsonData.fromJsonArrayString(
              '[{"hello": "world", "list": [1, null, true, "flutter", false, [456, 654], {"deep-level": null, "deeper-level": {"hello": "world"}}]}]'),
          [
            {
              'hello': 'world',
              'list': [
                1,
                null,
                true,
                'flutter',
                false,
                [456, 654],
                {
                  'deep-level': null,
                  'deeper-level': {'hello': 'world'}
                }
              ]
            }
          ]);
    });
  });
}


lib/json_data.dart
library json_data;

import 'dart:convert';

class JsonData {
  JsonData();

  static dynamic _convertValue(dynamic value,
      {bool wrapStringWithDoubleQuote = true}) {
    if (value is String) {
      if (wrapStringWithDoubleQuote) {
        if (value.startsWith('"') && value.endsWith('"')) {
          return value;
        } else {
          return '"$value"';
        }
      } else {
        if (value.startsWith('"') && value.endsWith('"')) {
          return value.substring(1, value.length);
        } else {
          return value.toString();
        }
      }
    } else if (value is List) {
      return _toJsonArray(value,
          wrapStringWithDoubleQuote: wrapStringWithDoubleQuote);
    } else if (value is Map) {
      return _toJson(value,
          wrapStringWithDoubleQuote: wrapStringWithDoubleQuote);
    } else if (value is bool) {
      return value;
    } else if (value == null) {
      return null;
    }
    return value;
  }

  static _toJson(Map<dynamic, dynamic> data,
      {bool wrapStringWithDoubleQuote = false}) {
    return data.map((key, value) => MapEntry(
        _convertValue(key,
            wrapStringWithDoubleQuote: wrapStringWithDoubleQuote),
        _convertValue(value,
            wrapStringWithDoubleQuote: wrapStringWithDoubleQuote)));
  }

  static String toJsonString(Map<String, Object?> data,
      {bool wrapStringWithDoubleQuote = true}) {
    return data
        .map((key, value) => MapEntry(
            _convertValue(key,
                wrapStringWithDoubleQuote: wrapStringWithDoubleQuote),
            _convertValue(value,
                wrapStringWithDoubleQuote: wrapStringWithDoubleQuote)))
        .toString();
  }

  static _toJsonArray(List<dynamic> dataList,
      {bool wrapStringWithDoubleQuote = false}) {
    return dataList
        .map((data) => _convertValue(data,
            wrapStringWithDoubleQuote: wrapStringWithDoubleQuote))
        .toList();
  }

  static String toJsonArrayString(List<dynamic> dataList,
      {bool wrapStringWithDoubleQuote = true}) {
    return dataList
        .map((data) => _convertValue(data,
            wrapStringWithDoubleQuote: wrapStringWithDoubleQuote))
        .toList()
        .toString();
  }

  static Map<String, dynamic> fromJsonString(String source) {
    var result = jsonDecode(source) as Map<String, dynamic>;
    for (var key in result.keys) {
      if (result[key] is String) {
        if (result[key].toString().startsWith('[') &&
            result[key].toString().endsWith(']')) {
          result[key] = JsonData.fromJsonArrayString(result[key].toString());
        } else if (result[key].toString().startsWith('{') &&
            result[key].toString().endsWith('}')) {
          result[key] = JsonData.fromJsonString(result[key]);
        } else {
          result[key] = JsonData._convertValue(result[key],
              wrapStringWithDoubleQuote: false);
        }
      } else {
        result[key] = JsonData._convertValue(result[key],
            wrapStringWithDoubleQuote: false);
      }
    }
    return Map<String, dynamic>.from(result);
  }

  static List<dynamic> fromJsonArrayString(String source) {
    var result = jsonDecode(source) as List<dynamic>;
    for (var i = 0; i < result.length; i++) {
      if (result[i] is String) {
        if (result[i].toString().startsWith('[') &&
            result[i].toString().endsWith(']')) {
          result[i] = JsonData.fromJsonArrayString(result[i].toString());
        } else if (result[i].toString().startsWith('{') &&
            result[i].toString().endsWith('}')) {
          result[i] = JsonData.fromJsonString(result[i].toString());
        } else {
          result[i] = JsonData._convertValue(result[i],
              wrapStringWithDoubleQuote: false);
        }
      } else {
        result[i] =
            JsonData._convertValue(result[i], wrapStringWithDoubleQuote: false);
      }
    }
    return List<dynamic>.from(result);
  }
}


example/lib/main.dart
import 'package:json_data/json_data.dart';

class Example {
  testToJsonString() {
    print(JsonData.toJsonString({
      'hello': 'world',
      'list': [
        1,
        null,
        true,
        'flutter',
        false,
        [456, 654],
        {
          'deep-level': null,
          'deeper-level': {'hello': 'world'}
        }
      ]
    }));
  }

  testToJsonArrayString() {
    print(JsonData.toJsonArrayString([
      {
        'hello': 'world',
        'list': [
          1,
          null,
          true,
          'flutter',
          false,
          [456, 654],
          {
            'deep-level': null,
            'deeper-level': {'hello': 'world'}
          }
        ]
      }
    ]));
  }

  testFromJsonString() {
    print(JsonData.fromJsonString(
        '{"hello": "world", "list": [1, null, true, "flutter", false, [456, 654], {"deep-level": null, "deeper-level": {"hello": "world"}}]}'));
  }

  testFromJsonArrayString() {
    print(JsonData.fromJsonArrayString(
        '[{"hello": "world", "list": [1, null, true, "flutter", false, [456, 654], {"deep-level": null, "deeper-level": {"hello": "world"}}]}]'));
  }
}
